Source/WebKit/UIProcess/Launcher/glib/ProcessLauncherGLib.cpp

```diff
@@ -235,21 +235,37 @@ void ProcessLauncher::launchProcess()
         g_error("Unable to spawn a new child process: %s", error->message);
 
 #if OS(LINUX)
-    m_processID = IPC::readPIDFromPeer(pidSocketPair.server);
-    RELEASE_ASSERT(!close(pidSocketPair.server));
+    GRefPtr<GSocket> pidSocket = adoptGRef(g_socket_new_from_fd(pidSocketPair.server, &error.outPtr()));
+    if (!pidSocket)
+        g_error("Failed to create pid socket wrapper: %s", error->message);
+
+    // We need to get the pid of the actual WebKit auxiliary process, not the bwrap or flatpak-spawn
+    // intermediate process. And do it without blocking, because process launching is slow.
+    g_socket_set_blocking(pidSocket.get(), FALSE);
+    m_socketMonitor.start(pidSocket.get(), G_IO_IN, RunLoop::main(), [protectedThis = Ref { *this }, this, pidSocket, serverSocket = webkitSocketPair.server](GIOCondition condition) -> gboolean {
+        if (condition & G_IO_ERR || condition & G_IO_NVAL)
+            g_error("Failed to read pid from child process");
```

Comment:
We want to handle `G_IO_HUP` here as well to detect whether the socket pair has been closed prematurely on the other end. We may want to add `safeStrerror(errno)` to the error message as well, in case it may help troubleshoot issues, may they arise.

Reply:
I actually tried that, but it didn't work. The problem is the child process is going to send its message and then immediately close the socket, so `G_IO_HUP` is one of the events that gets received here normally.

That said, maybe I should detect `(condition & G_IO_HUP) && !(condition & G_IO_IN)` as an error. That shouldn't ever happen, but if it does, the ProcessLauncher will just get stuck forever without failing.

> We may want to add safeStrerror(errno) to the error message as well, in case it may help troubleshoot issues, may they arise.

Hm, I'm not sure, because we're not directly calling any function that sets errno.

Reply:
> That said, maybe I should detect (condition & G_IO_HUP) && !(condition & G_IO_IN) as an error. That shouldn't ever happen, but if it does, the ProcessLauncher will just get stuck forever without failing.

(I'm not sure whether this can actually happen or not.)

Reply:
> That said, maybe I should detect (condition & G_IO_HUP) && !(condition & G_IO_IN) as an error. That shouldn't ever happen, but if it does, the ProcessLauncher will just get stuck forever without failing.

Looking over the `poll(2)` man page, I think this is correct. For our purposes, it's an error only if it occurs on its own.

Reply:
Updated. This look OK? Please assign to merge queue if so.

Reply:
Maybe it would be simpler to first check `condition & G_IO_IN` and then you only need to check the others after the return